### DISCUSSION TITLE
Extractors for parts of state that are under a Mutex

### DISCUSSION THREAD

**Discussion by robryk** [tokio-rs/axum#2914] [Q&A]
### Summary

I think I have another usecase for [explicit 'state lifetime in FromRequestParts](https://github.com/tokio-rs/axum/issues/1374#issuecomment-1250365505) and would like to know if there's another way of achieving the same.

I have a URL pattern that contains a resource ID. My state is morally equivalent to `Arc<Mutex<HashMap<String, Resource>>>` with the `String` corresponding to the resource ID. I would like to have an extractor that extracts "access" to the appropriate `Resource`.

I can make a wrapper `struct ResourceHandle<'a>(&'a mut MappedMutexGuard<'a, Resource>)` and could give it a `FromRequestParts<my_state>` implementation, except that I cannot guarantee that the state will outlive `'a`.

An obvious alternative is to replace `Resource` with `Arc<Mutex<Resource>>` in my state, and have an extractor for that (which will simply clone the Arc). That is certainly doable (and allows for more concurrency -- the global mutex is held only for the duration of the lookup), but it's somewhat more cumbersome to deal with (each handler has to deal with locking the per-resource mutex).

Is there a way of doing some moral equivalent of the former?

### axum version

0.7.5

**Comment by jplatte** [comment]
If you have a tokio `Mutex` and want to lock it through an extractor until the handler returns or the guard (extractor) is explicitly dropped, have a look at [`Mutex::lock_owned`](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.lock_owned). Does that help?
